src/scan/postgres_table_reader.cpp

```diff
@@ -0,0 +1,177 @@
+		memcpy(parallel_worker_readers, pei->reader, nreaders * sizeof(TupleQueueReader *));
+	}
+
+	HOLD_CANCEL_INTERRUPTS();
+
+	slot = ExecInitExtraTupleSlot(table_scan_query_desc->estate, table_scan_planstate->ps_ResultTupleDesc, &TTSOpsMinimalTuple);
+
+	restore_stack_base(current_stack);
+}
+
+PostgresTableReader::~PostgresTableReader() {
+	std::lock_guard<std::mutex> lock(DuckdbProcessLock::GetLock());
+
+	ExecEndNode(table_scan_planstate);
+
+	if (pei != NULL)
+		ExecParallelFinish(pei);
+
+	if (parallel_worker_readers) {
+		pfree(parallel_worker_readers);
+	}
+
+	parallel_worker_readers = NULL;
+
+	ExecParallelCleanup(pei);
+
+	pei = NULL;
+
+	ExecutorFinish(table_scan_query_desc);
+	ExecutorEnd(table_scan_query_desc);
+	FreeQueryDesc(table_scan_query_desc);
+}
+
+/*
+ * This is simple calculation how much postgresql workers we will wanted to spawn for this
+ * postgres table scan.
+ */
+int
+PostgresTableReader::ParalleWorkerNumber(Cardinality cardinality) {
+	static const int base_log = 8;
+	int cardinality_log = std::log2(cardinality);
+	int base = cardinality_log / base_log;
+	return std::max(1, std::min(base, std::min(max_workers_per_postgres_scan, max_parallel_workers)));
+}
+
+TupleTableSlot *
+PostgresTableReader::GetNextTuple() {
+	MinimalTuple worker_minmal_tuple;
+	if (nreaders > 0) {
+		worker_minmal_tuple = GetNextWorkerTuple();
+		if (HeapTupleIsValid(worker_minmal_tuple)) {
+			ExecStoreMinimalTuple(worker_minmal_tuple, slot, false);
+			return slot;
+		}
+	} else {
+		std::lock_guard<std::mutex> lock(DuckdbProcessLock::GetLock());
+		pg_stack_base_t current_stack = set_stack_base();
+		table_scan_query_desc->estate->es_query_dsa = pei ? pei->area : NULL;
+		slot = ExecProcNode(table_scan_planstate);
+		restore_stack_base(current_stack);
+		table_scan_query_desc->estate->es_query_dsa = NULL;
+		if (!TupIsNull(slot)) {
+			return slot;
+		}
+	}
+
+	return ExecClearTuple(slot);
+}
+
+MinimalTuple
+PostgresTableReader::GetNextWorkerTuple() {
+	int nvisited = 0;
+	for (;;) {
+		TupleQueueReader *reader;
+		MinimalTuple minimal_tuple;
+		bool readerdone;
+
+		reader = (TupleQueueReader *) parallel_worker_readers[next_parallel_reader];
+		minimal_tuple = TupleQueueReaderNext(reader, true, &readerdone);
+
+		if (readerdone) {
+			--nreaders;
+			if (nreaders == 0) {
+				return NULL;
+			}
+			memmove(&parallel_worker_readers[next_parallel_reader], &parallel_worker_readers[next_parallel_reader + 1],
+			        sizeof(TupleQueueReader *) * (nreaders - next_parallel_reader));
+			if (next_parallel_reader >= nreaders) {
+				next_parallel_reader = 0;
+			}
+			continue;
+		}
+
+		if (minimal_tuple) {
+			return minimal_tuple;
+		}
+
+		next_parallel_reader++;
+		if (next_parallel_reader >= nreaders)
+			next_parallel_reader = 0;
+
+		nvisited++;
+		if (nvisited >= nreaders) {
+			(void)WaitLatch(MyLatch, WL_LATCH_SET | WL_EXIT_ON_PM_DEATH, 0, WAIT_EVENT_EXECUTE_GATHER);
+			ResetLatch(MyLatch);
```

Comment:
Calling these functions requires the global lock, and it doesn't seem like we're holding it here.